graphql/resolve/mutation_rewriter.go

```diff
@@ -1375,44 +1399,79 @@ func rewriteObject(
 				xidString, _ = extractVal(xidVal, xid.Name(), xid.Type().Name())
 				variable = varGen.Next(typ, xid.Name(), xidString, false)
 
-				// Three cases:
-				// 1. If the queryResult UID exists. Add a reference.
+				// If this xid field is inherited from interface and have interface argument set, we also
+				// have existence query for interface to make sure that this xid is unique across all
+				// implementation types of the interface.
+				// We have following cases
+				// 1. If the queryResult UID exists for any of existence query (type or interface),
+				//    then add a reference.
 				// 2. If the queryResult UID does not exist and this is the first time we are seeing
 				//    this. Then, return error.
 				// 3. The queryResult UID does not exist. But, this could be a reference to an XID
 				//    node added during the mutation rewriting. This is handled by adding the new blank UID
 				//    to existenceQueryResult.
 
-				// Get whether node with XID exists or not from existenceQueriesResult
-				if uid, ok := idExistence[variable]; ok {
+				interfaceTypDef, interfaceVar := interfaceVariable(typ, varGen, xid.Name(), xidString)
+
+				// Get whether node with XID exists or not from existenceQueriesResults
+				_, interfaceUidExist := idExistence[interfaceVar]
+				typUid, typUidExist := idExistence[variable]
+
+				if interfaceUidExist || typUidExist {
 					// node with XID exists. This is a reference.
```

Comment:
We are making the assumption in the algorithm that if typeUidExist is true then interfaceUidExist is also true. This assumption is valid for newly generated schemas but may not be valid from schemas edited after inserting data.
Can you add the following assert:
if typeUidExists {
  // Assert that interfaceUidExists is also true.
}